[Geo] Decouple geojson parse logic from ShapeBuilders

nknize opened the pull request:
This is the first step to supporting WKT (and other future) format(s). The ShapeBuilders are still quite messy but are further simplified by decoupling the parse logic from the build logic. This PR refactors the parsing logic into its own package separate from the Shape builders. It also decouples the GeoShapeType into a standalone enumerator that is responsible for validating the parsed data and providing the appropriate builder. This future-proofs the code making it easier to maintain and not only add new shape types but support for new geo formats (e.g., WKT).

cbuescher commented on core/src/main/java/org/elasticsearch/common/geo/builders/CoordinateCollection.java:
```diff
@@ -21,24 +21,31 @@
  * It holds a common list of {@link Coordinate}, provides setters for adding elements to the list and can render this to XContent.
  */
-public abstract class CoordinateCollection<E extends CoordinateCollection<E>> extends ShapeBuilder {
+public abstract class CoordinateCollection<E extends CoordinateCollection<E>> implements NamedWriteable, ToXContentObject {
```
What is the reason this is moving up in the class hierarchy? I see that this used to be a supertype of LineStringBuilder and MultiPointBuilder, maybe to group common functionality. Does CoordinateCollection provide that functionality to all subclasses now? Can we maybe merge ShapeBuilder and CoordinateCollection now? I see that all ShapeBuilders are now also CoordinateCollections, maybe we should only keep ShapeBuilders?

nknize replied:
Yeah, merging would probably be the best idea to eliminate superfluous classes. I'll have a look at making that change.

cbuescher commented on core/src/main/java/org/elasticsearch/common/geo/parsers/ShapeParser.java:
```diff
@@ -0,0 +1,128 @@
+     * Can either be a leaf node consisting of a Coordinate, or a parent with
+     * children
+     */
+    public static class CoordinateNode implements ToXContentObject {
```
Does it make sense to move this to its own class? If not, can access be limited so it can only be used in this package?

nknize replied:
I can split this into its own class under the `parsers` package. Its unique to parsing but also reference by `GeoShapeType`

cbuescher commented on client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/QueryDSLDocumentationTests.java:
```diff
@@ -189,7 +189,7 @@ public void testGeoShape() throws IOException {
             GeoShapeQueryBuilder qb = geoShapeQuery(
                     "pin.location",                                      // <1>
-                    ShapeBuilders.newMultiPoint(                         // <2>
+                    new MultiPointBuilder(                         // <2>
```
I don't think we can remove the ShapeBuilders helper class in 6.1., that might break some peoples code and I think this is user-facing enough. Not sure if that was the plan, I think its okay to break in 7.0 but then it needs to be deprecated.

nknize replied:
:+1:  I can keep this PR for 7.0 and open a separate deprecation PR for 6.1?

cbuescher replied:
Sound good.

cbuescher commented on core/src/main/java/org/elasticsearch/common/geo/builders/ShapeBuilders.java:
```diff
@@ -1,153 +0,0 @@
-     * Create a new set of points
-     * @return new {@link MultiPointBuilder}
-     */
-    public static MultiPointBuilder newMultiPoint(List<Coordinate> points) {
```
I think we need to keep these methods at least in 6.1

nknize replied:
Will do... I'll handle that in a 6.1 PR.

cbuescher commented on core/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java:
```diff
@@ -445,27 +447,41 @@ public Mapper parse(ParseContext context) throws IOException {
-                throw new MapperParsingException("[{" + fieldType().name() + "}] is configured for points only but a " +
+            if (fieldType().pointsOnly() == true) {
+                // index configured for pointsOnly
+                if (shape instanceof XShapeCollection && XShapeCollection.class.cast(shape).pointsOnly()) {
```
Just out of curiosity, where does this piece of code come from? Is this merely a refactoring or new behaviour?

nknize replied:
Its technically a bug fix. Indexing would fail if `ShapeCollection` was a collection of `Point` types. That shouldn't happen on an index configured for `points_only : true`

cbuescher replied:
Since this already seems to be its own selfcontained commit, would if be difficult to pull it out into a separate PR then? Better to get its own comment etc. then.

nknize replied:
Yes, I like that idea. I'll open a separate PR and probably a separate issue for tracking purposes.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java:
```diff
@@ -410,7 +411,7 @@ public void onResponse(GetResponse response) {
                                 if (pathElements[currentPathSlot].equals(parser.currentName())) {
                                     parser.nextToken();
                                     if (++currentPathSlot == pathElements.length) {
-                                        listener.onResponse(ShapeBuilder.parse(parser));
```
Maybe we need to keep ShapeBuilder.parse() and forward to the new method in 6.1 (and deprecate).

nknize replied:
Yeah.. I can keep `ShapeBuilder.parse` in 6.1 and handle deprecation. I'll do that in another PR.

cbuescher commented on core/src/test/java/org/elasticsearch/common/geo/builders/MultiLineStringBuilderTests.java:
```diff
@@ -68,6 +68,6 @@ static MultiLineStringBuilder mutate(MultiLineStringBuilder original) throws IOE
 
     static MultiLineStringBuilder createRandomShape() {
-        return new MultiLineStringBuilder();
+        return MultiLineStringBuilder.class.cast(RandomShapeGenerator.createShape(random(), ShapeType.MULTILINESTRING));
```
Interesting syntax, any reason not to cast using "(MultiLineStringBuilder)" here? I'm just curious ;-)

nknize replied:
I used this syntax because the `cast` method call explicitly calls `instanceof`. I'm not sure if compiler optimization does that for me with the cast operator? Maybe so, at which point this is just added verbosity?

cbuescher replied:
+1

nknize commented on core/src/main/java/org/elasticsearch/common/geo/GeoShapeType.java:
```diff
@@ -0,0 +1,306 @@
+        }
+
+        @Override
+        protected CoordinateNode validate(CoordinateNode coordinates, boolean coerce) {
```
:+1:

nknize commented on core/src/main/java/org/elasticsearch/common/geo/GeoShapeType.java:
```diff
@@ -0,0 +1,306 @@
+
+    public static GeoShapeType forName(String geoshapename) {
+        String typename = geoshapename.toLowerCase(Locale.ROOT);
+        for (GeoShapeType type : values()) {
```
:+1:

nknize commented on core/src/main/java/org/elasticsearch/common/geo/GeoShapeType.java:
```diff
@@ -0,0 +1,306 @@
+
+    public abstract ShapeBuilder getBuilder(CoordinateNode coordinates, DistanceUnit.Distance radius,
+                                            ShapeBuilder.Orientation orientation, boolean coerce);
+    protected abstract CoordinateNode validate(CoordinateNode coordinates, boolean coerce);
```
:+1:

nknize commented on core/src/main/java/org/elasticsearch/common/geo/GeoShapeType.java:
```diff
@@ -0,0 +1,306 @@
+                                            ShapeBuilder.Orientation orientation, boolean coerce);
+    protected abstract CoordinateNode validate(CoordinateNode coordinates, boolean coerce);
+
+    public static void registerShapes(List<Entry> namedWriteables) {
```
:+1:

nknize commented on core/src/main/java/org/elasticsearch/common/geo/builders/CircleBuilder.java:
```diff
@@ -31,7 +33,7 @@
-public class CircleBuilder extends ShapeBuilder {
+public class CircleBuilder extends ShapeBuilder<Circle> {
 
     public static final String FIELD_RADIUS = "radius";
```
Just missed this one... I can, and probably should, do it now.

nknize commented on core/src/main/java/org/elasticsearch/common/geo/parsers/ShapeParser.java:
```diff
@@ -0,0 +1,128 @@
+/**
+ * static base class for implementing a shape parser
+ */
+public abstract class ShapeParser {
```
There's a plan to add a WKT parser in the next pull request. This PR is setting up for that new parser.

cbuescher replied:
Okay, can you briefly explain the (maybe future) relationship of the ShapeParser and the above GeoJsonParser class? Currently they both seem to mostly consist of a "static ShapeBuilder parse(XContentParser parser, GeoShapeFieldMapper shapeMapper)" method, don't have any state themselves but ShapeParser calls GeoJsonParser. It would be useful to understand where this is going.

nknize replied:
Yeah the more I think about this the more I don't think there is a need to ever have state here. So really the only relationship is that `ShapeParser` calls the appropriate "strategy" based on initial parsing. e.g. the following rough code:
```
public static ShapeBuilder parse(XContentParser parser, GeoShapeFieldMapper shapeMapper) throws IOException {
        if (parser.currentToken() == XContentParser.Token.VALUE_NULL) {
            return null;
        } else if (parser.currentToken() == XContentParser.Token.START_OBJECT) {
            return GeoJsonParser.parse(parser, shapeMapper);
        } else if (parser.currentTocken() == XContentParser.VALUE_STRING) {
            return GeoWKTParser.parse(parser, shapeMapper);
        }
        throw new ElasticsearchParseException("shape must be an object consisting of type and coordinates");
    }
```

cbuescher replied:
Thanks, I get the general idea now.

nknize commented:
Thanks for the initial review @cbuescher I added a commit for the first round updates to let jenkins run the test. I also removed the 6.1 label and will open a separate PR for deprecation. Once the jenkins tests pass I think this is close.

nknize commented:
@cbuescher looks like this is good to go pending a decision on `ShapeParser` and `GeoJsonParser` I could merge the `GeoJsonParser` logic in with `ShapeParser`, but then when I add the `GeoWKTParser` logic I'd be adding that to `ShapeParser` as well. The only justification for keeping them in their own class is for maintenance and code organization and preventing `ShapeParser` from getting huge. A good justification I think? How about you?
